### PR TITLE
dml: fix index column offset beyond the range

### PR DESCRIPTION
**Pull request by AilinKid**
Signed-off-by: ailinkid <314806019@qq.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/25342

### What is changed and how it works?

Make the DELETE statement aware of middle state column, while user can't see them yet (non-explicit usage).
(some indexes convering the dropping-column need check the middle state column to remove the index record)

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- dml: fix index column offset beyond the range

### CODE REVIEW

**Review by wjhuang2016 — COMMENTED** [review]
It seems that this PR will revert #17048

**Comment by AilinKid** [comment]
> It seems that this PR will revert #17048

kind of, 17048's target test has been works under the latter fix of jiwei's use of `NotExplicitUsable`, using the deletableCol is the better choice.

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -3903,7 +3903,7 @@ func (b *PlanBuilder) buildDataSource(ctx context.Context, tn *ast.TableName, as
 		columns = tbl.WritableCols()
 	} else if b.inDeleteStmt {
 		// All hidden columns are needed because we need to delete the expression index that consists of hidden columns.
```
Comment (zimulala) [inline planner/core/logical_plan_builder.go:3905]: We need to update this comment.

**Review by zimulala — COMMENTED** [review]
So could we remove `FullHiddenColsAndVisibleCols`?

File: planner/core/logical_plan_builder.go (hunk 1 above)
Comment (AilinKid) [reply planner/core/logical_plan_builder.go]: FullHiddenColsAndVisibleCols() is useful and depended in the case of partion prune, so I keeped it.

```
func (s *partitionProcessor) reconstructTableColNames(ds *DataSource) ([]*types.FieldName, error) {
	names := make([]*types.FieldName, 0, len(ds.TblCols))
	colsInfo := ds.table.FullHiddenColsAndVisibleCols()
	colsInfoMap := make(map[int64]*table.Column, len(colsInfo))
	for _, c := range colsInfo {
		colsInfoMap[c.ID] = c
	}
        ...
```

**Review by wjhuang2016 — APPROVED** [review]

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**